sklearn/metrics/_safe_euclidean_sparse.pyx

```diff
@@ -0,0 +1,69 @@
+#cython: language_level=3
+#cython: boundscheck=False, cdivision=True, wraparound=False
+
+
+import numpy as np
+cimport numpy as np
+from cython cimport floating
+from libc.math cimport fmax
+
+
+np.import_array()
+
+
+ctypedef fused INT:
+    np.int32_t
+    np.int64_t
+
+
+def _euclidean_sparse_dense_exact(floating[::1] X_data,
+                                  INT[::1] X_indices,
+                                  INT[::1] X_indptr,
+                                  np.ndarray[floating, ndim=2, mode='c'] Y,
+                                  floating[::1] y_squared_norms):
+    cdef:
+        int n_samples_X = X_indptr.shape[0] - 1
+        int n_samples_Y = Y.shape[0]
+        int n_features = Y.shape[1]
+
+        int i, j
+
+        floating[:, ::1] D = np.empty((n_samples_X, n_samples_Y), Y.dtype)
+
+    for i in range(n_samples_X):
+        for j in range(n_samples_Y):
+            D[i, j] = _euclidean_sparse_dense_exact_1d(
+                &X_data[X_indptr[i]],
+                &X_indices[X_indptr[i]],
+                X_indptr[i + 1] - X_indptr[i],
+                &Y[j, 0],
+                y_squared_norms[j])
+
+    return np.asarray(D)
+
+
+cdef floating _euclidean_sparse_dense_exact_1d(floating *y_data,
+                                               INT *y_indices,
+                                               int y_nnz,
+                                               floating *x,
+                                               floating x_squared_norm) nogil:
+    """Euclidean distance between x dense and y sparse"""
+    cdef:
+        int i
+        floating xi
+        floating tmp = 0.0
+        floating result = 0.0
+        floating partial_x_squared_norm = 0.0
+    
+    # Split the loop to avoid unsafe compiler auto optimizations
```

Comment:
I might say something stupid, but I'm not sure what you're trying to avoid there.
If a compiler optimize agressively enough to try to rearrange your arithmetic operations in an unsafe way, then I don't think fusing two independant loops would be a problem for it.
Additionally, AFAIK, most (if not all) compilers nowadays know pretty well how floating point arithmetic work.

Reply:
I was also wondering about this -- I imagine, there should be no unsafe optimizations unless you are using non-standard flags (e.g. `-Ofast`)?

Reply:
I thought that too, but this is not what I observe. I don't really know what's going on. The loop should be:
```python
for i in range(y_nnz):
    xi = x[y_indices[i]]
    tmp = y_data[i] - xi
    result += (tmp * tmp) - (xi * xi)
```
but even without the `-Ofast` flag, there's is some kind of optimization because the result is not correct.
It looks like gcc expands the expression
```
result += (tmp * tmp) - (xi * xi) = (y_data[i] - xi)² -xi²
                                  = y_data[i]² - 2*y_data[i]*xi
```
which might be wrong in floating point arithmetic.

Reply:
Make sure to add a unit test that detects bad numeric optimization!

Reply:
how would you do that ?
I added tests for the euclidean distances which failed when I didn't split the loop or when I use the -Ofast flag. Now they pass. Is that what you mean ?

Reply:
> but even without the `-Ofast` flag, there's is some kind of optimization because the result is not correct.

With gcc, make sure you don't pass `-ffast-math` directly or indirectly. Maybe try playing around with `gcc -Q --help=optimizers` and check what's enabled.

> which might be wrong in floating point arithmetic.

`tmp² - xi²` already has a significant loss of accuracy when `tmp` is close to `xi`. However, inlining the subtraction of `xi²` should reduce the maximal absolute value reached by `result` compared to waiting until after the second loop to subtract the whole sum. Thus resulting in a better accuracy.

Also, note that `(tmp * tmp) - (xi * xi)` can be rewritten as `(tmp - xi) * (tmp + xi)` which should have a better accuracy at the expense of an additional operation.
It can be further simplified into `(y_data[i] - xi - xi) * (y_data[i] - xi + xi) = (y_data[i] - 2*xi) * y_data[i]`, which might save one more bit of accuracy.
Note that it's very close to what gcc did.

Of course, the source of this formula should be documented as it's really not obvious that it computes an euclidean distance.

Reply:
I don't pass -ffast-math. The only flag cython passes is -O3 which does not include fast-math.

Reply:
none of the 2 other formulations work. I though the `(tmp - xi) * (tmp + xi)` formulation would give good enough precision...
I'm not an expert of floating point arithmetic, I don't really get what's happening...

Here's is the detail of the formula for the distance computation (remind that x is sparse)
![CodeCogsEqn1](https://user-images.githubusercontent.com/34657725/54359557-349e3180-4663-11e9-8900-e341c777d711.gif)
![CodeCogsEqn2](https://user-images.githubusercontent.com/34657725/54359567-3a941280-4663-11e9-966b-2f2ad677e02d.gif)
up to here, there shouldn't be any precision loss. It's still the exact method, with the sum split.
It's what's implemented in this PR. The 3rd term is the first loop. The first term is the second loop and the last term is the norm added at the end.

Now, we can regroup the 2 loops into:
![CodeCogsEqn3](https://user-images.githubusercontent.com/34657725/54359574-3cf66c80-4663-11e9-8d42-c0d995eeae1c.gif)
but this causes loss of precision

Reply:
If x is close to y, the first term will be close to 0, and the result of d(x,y) will be based on the difference of "almost the norm of y" and the norm of y, at which point the precision is lost because you subtract close values.
Compared to these big values, the actual signal (xi-yi)^2 is small, and gets lost in precision.
Unfortunately, this happens in particular for near neighbors, where we usually need high precision.

Reply:
I'm not sure which part you're talking about. Equations 1 and 2 can't have precision loss. It's just the full some split in several terms. This is what I implemented.

Reply:
Parts (terms) of the last equation.

Reply:
A similar loss likely happens here: `(tmp * tmp) - (xi * xi)`, when `tmp` and `xi` are of similar magnitude this equations does have catastrophic cancellation (here this happens when yi is small).

Reply:
> I'm not an expert of floating point arithmetic

Neither am I, so take everything I say here with a grain of salt.

First, I'd like to make a note of the vocabulary around floating point arithmetic.

- *precision* refer to the number of bits used to make the calculation.
- *accuracy* refer to how close to the real value a result is.
- *error* refer to the difference between the real value and what's actually stored.

> up to here, there shouldn't be any precision loss.

Well, none your transformations are exactly equivalent. Floating point arithmetic is commutative, but not associative. In general `(a+b)+c != a+(b+c)`. You can't reorder a sum as you wish without changing the result. That might improve the accuracy, but that might also make it worse. You usually want to add values that are not too far from each other in order to avoid cancellation.
As a side note about summation, numpy implement a [https://en.wikipedia.org/wiki/Pairwise_summation](pairwise summation) which has a better accuracy than the straightforward sum.

> Now, we can regroup the 2 loops into:
> but this causes loss of precision

What makes you say so? How did you measure it?

Reply:
> What makes you say so? How did you measure it?

well just that the tests fails when I regroup the 2 loops :)

Reply:
> well just that the tests fails when I regroup the 2 loops :)

Which test? I can't reproduce.